### PR TITLE
[DEV-9552] Current CD - Advanced Search

### PR DESCRIPTION
**Pull request by dporth-frb**
**Description:**
This PR introduces two new location filter terms, `district_original` and `district_current`. `district_original` is replacing the existing location filter term `district`. `district` is deprecated now and expected to be fully deleted in the future. `district_current` is a brand new concept and now available to be used in `place_of_performance` and `recipient_locations` location objects.

**Technical details:**
Another aspect of this PR is that it fixed an unrelated, production level bug in the `spending_by_geography` endpoint. Please see my comment making known where this bug occurred and how it was fixed. Fixing this bug was necessary in order to satisfy the AC of this story.

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
2. [x] API documentation updated
3. [x] Necessary PR reviewers:
    - [x] Backend
    - [x] Frontend <OPTIONAL>
    - [x] Operations <OPTIONAL>
    - [x] Domain Expert <OPTIONAL>
4. [x] Matview impact assessment completed (N/A)
5. [x] Frontend impact assessment completed (N/A)
6. [x] Data validation completed
7. [x] Appropriate Operations ticket(s) created (N/A)
8. [x] Jira Ticket [DEV-9552](https://federal-spending-transparency.atlassian.net/browse/DEV-9552):
    - [x] Link to this Pull-Request
    - [x] Performance evaluation of affected (API | Script | Download) (N/A)
    - [x] Before / After data comparison

**Area for explaining above N/A when needed:**
4. The nature of this work does not impact materialized views
5. The nature of this work does not require frontend impact assessments
7. The nature of this work does not require operation tickets.
8b. The nature of this work does not impact performance

### CODE REVIEW

File: usaspending_api/search/v2/views/spending_by_geography.py [hunk 1]
```diff
@@ -287,7 +299,7 @@ def county_district_queryset(
                 {"state": fips_to_code.get(x[:2]), self.geo_layer.value: x[2:], "country": "USA"}
                 for x in self.geo_layer_filters
             ]
-            self.queryset = self.queryset.filter(geocode_filter_locations(scope_field_name, geo_layers_list))
```
Comment (dporth-frb) [inline usaspending_api/search/v2/views/spending_by_geography.py:290]: This change is the production level bug that I had to fix, as mentioned in my PR description technical notes.

Comment (dporth-frb) [reply usaspending_api/search/v2/views/spending_by_geography.py:290]: The following shows how to reproduce this bug using the state of the production API as it is today.
```
POST /api/v2/search/spending_by_geography/ HTTP/1.1
Host: api.usaspending.gov
Content-Type: application/json
Cookie: cookiesession1=678A3E0DE0F2E39CB6449AE9B332E675
Content-Length: 363

{
    "scope": "recipient_location",
    "geo_layer": "district",
    "geo_layer_filters": [
        "OR"
    ],
    "filters": {
        "time_period": [
            {
                "start_date": "2007-10-01",
                "end_date": "2023-09-30"
            }
        ]
    },
    "subawards": true,
    "auditTrail": "Map Visualization"
}
```

File: usaspending_api/api_contracts/search_filters.md [hunk 2]
```diff
@@ -113,9 +113,15 @@ Keys in a location object include:
...
-    * If `district` is provided, a `county` *must never* be provided.
+* **district** - Deprecated. Use `district_original`.
+* **district_original** - a 2 character code indicating the congressional district
+    * When provided, a `state` must always be provided as well.
+    * When provided, a `county` *must never* be provided.
+    * When provided, `country` must always be "USA".
+* **district_current** - a 2 character code indicating the current congressional district
+    * When provided, a `state` must always be provided as well.
+    * When provided, a `county` *must never* be provided.
+    * When provided, `country` must always be "USA".
```
Comment (collinwr) [inline usaspending_api/api_contracts/search_filters.md:124]: Might want to make a note that when a `district_current` is provided, a `district_original` must not be provided, and vice versa

File: usaspending_api/awards/v2/filters/location_filter_geocode.py [hunk 3]
```diff
@@ -57,6 +58,15 @@ def geocode_filter_locations(scope: str, values: list) -> Q:
...
     for v in values:
+        state = v.get("state")
+        country = v.get("country")
+        county = v.get("county")
+        district_current = v.get("district_current")
+        district_original = v.get("district_original")
+        if (state is None or country != "USA" or county is not None) and (
+            district_current is not None or district_original is not None
+        ):
+            raise InvalidParameterException(INCOMPATIBLE_DISTRICT_LOCATION_PARAMETERS)
```
Comment (collinwr) [inline usaspending_api/awards/v2/filters/location_filter_geocode.py:69]: Same as comment about markdown updates. Probably want to ensure that `district_original` and `district_current` aren't included in the same Location Object

File: usaspending_api/api_contracts/search_filters.md (hunk 2 above)
Comment (dporth-frb) [reply usaspending_api/api_contracts/search_filters.md:124]: Good point. Will do.

Comment (dporth-frb) [reply usaspending_api/api_contracts/search_filters.md:125]: Fixed

**Review by collinwr — DISMISSED** [review]

**Review by collinwr — APPROVED** [review]

**Pull request merged**